scipy/stats/_sensitivity_analysis.py

```diff
@@ -0,0 +1,418 @@
+) -> Tuple[np.ndarray,  np.ndarray]:
+    r"""Saltelli2010 formulation.
+
+    .. math::
+
+        S_i = \frac{1}{N} \sum_{j=1}^N
+        f(\mathbf{B})_j (f(\mathbf{AB}^{(i)})_j - f(\mathbf{A})_j)
+
+    .. math::
+
+        S_{T_i} = \frac{1}{N} \sum_{j=1}^N
+        (f(\mathbf{A})_j - f(\mathbf{AB}^{(i)})_j)^2
+
+    Parameters
+    ----------
+    f_A, f_B, f_AB : array_like (n, x)
+        Function evaluations ``n`` being the number of evaluations and ``x``
+        a vector output.
+
+    Returns
+    -------
+    s, st : array_like (x, d)
+        First order and total order Sobol' indices.
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    f_AB = f_AB.reshape(-1, *f_A.shape)
+
+    var = np.var(np.vstack([f_A, f_B]), axis=0)
+
+    s = np.mean(f_B * (f_AB - f_A), axis=1) / var
+    st = 0.5 * np.mean((f_A - f_AB) ** 2, axis=1) / var
+
+    return s.T, st.T
+
+
+@dataclass
+class SobolResult:
+  first_order: np.ndarray
+  total_order: np.ndarray
+  bootstrap_result: BootstrapResult
+
+
+def sobol_indices(
+    *,
+    func: Callable,
+    n: IntNumber,
+    dists: List[PINVDist],
+    random_state: SeedType = None
+) -> SobolResult:
+    r"""Global sensitivity indices of Sobol'.
+
+    Parameters
+    ----------
+    func : callable
+        Function to compute the Sobol' indices from. It's signature must be
+        ``func(x: ArrayLike) -> ArrayLike``, with ``x`` of shape ``(-1, d)``
+        and the output should have a shape ``(-1, s)`` with ``s`` the number
+        of output.
+    n : int
+        Must be a power of 2. The total number of function call will be
+        ``n(d+2)``.
+    dists : list(distributions), optional
+        List of distributions of the parameters. It must be compatible with
+        `scipy.stats.sampling.NumericalInversePolynomial`.
+    random_state : {None, int, `numpy.random.Generator`}, optional
+        If `random_state` is an int or None, a new `numpy.random.Generator` is
+        created using ``np.random.default_rng(random_state)``.
+        If `random_state` is already a ``Generator`` instance, then the
+        provided instance is used.
+
+    Returns
+    -------
+    res : SobolResult
+        An object with attributes:
+
+        first_order : ndarray
+            ...
+        total_order : ndarray
+            ...
+        bootstrap_result : BootstrapResult
+            An object providing confidence intervals on the indices.
+            See `scipy.stats.bootstrap` for more details.
+
+            The bootstrapping is done on both first and total order indices
+            at the same time meaning both bootstrap is a concatenated array.
```

Comment:
`bootstrap` produces confidence intervals by default in the result object because that's _all_ if was originally designed to do. Every other new function that produces confidence intervals is doing so by calling a `confidence_interval` method of the result object. 

One reason is that computation of confidence intervals is often slower than the basic calculation, so it should be done only if requested by the user. The other - perhaps more important - is that computation of a confidence interval may require a totally new set of arguments that would be inelegant to pass into the main function; e.g. `confidence_level`, `method` (if there are different ways of estimating it), `n_resamples` (in the case of a bootstrapped interval).

If I could go back and change it, I might even make `bootstrap` return a result object with just the bootstrap distribution and `standard_error` as attributes; it would have a `confidence_interval` method, too. I can't change that now, but we can ensure that all _new_ functions have the same interface.

In this case, maybe there should just be a `bootstrap` method instead of a `confidence_interval` method, since it returns a full `bootstrap` result?

Reply:
Yes.